### PR TITLE
Improve correctness of auto_parallel

### PR DESCRIPTION
**Pull request by sj6077**
This PR improves the correctness of current auto_parallel in grappler.

The problem of current auto_parallel can be summarized as below.
**First, each replica updates a variable independently.** This computation result could be different with a single device, which computes gradients for large batch size and updates a variable once if you use complex optimizers like momentum optimizer.  
**Second, the result of gradient clipping is changed.** In the current version, auto_parallel aggregates the input of variable update operators(e.g., ApplyGradientDescent) so "clipping->aggregation" is the current order if clipping is used. However, “aggregation->clipping” is the right order to make the same result as a single-device job.

This PR changes to aggregate gradients from auto differentiation of trainable variables.
The aggregations are defined explicitly instead of implicit aggregation through update operators. 

**[Current auto_parallel]**
current_auto_parallel

**[auto_parallel in this PR]**
auto_parallel_after_pr

**Changes in this PR**
1. Assumptions
Current: variables are updated by optimizer operations
This PR : all the gradients are made by tf.gradients, the gradients of trainable variables are the target of aggregation in synchronous training

2. Correct gradient clipping(e.g. tf.clip_by_global_norm, tf.clip_by_value)
Refer figures above.

3. The number of updates
Current : as many as the number of replicas(refer above)
This PR : one (it can also improve performance on multi-machine environment)

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @akshaym: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 44 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 59 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 74 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 89 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 104 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 119 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 134 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 149 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @rmlarsen, @ezhulenev: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 164 days with no activity and the `awaiting review` label has been applied.
